### PR TITLE
Flush data from the aggregator and forwarder when stopping the agent.

### PR DESCRIPTION
**Pull request by hush-hush**
### What does this PR do?

Flush data from the aggregator and forwarder when stopping the agent.

The aggregator will now flush aggregated metrics (closed time buckets and finish check's metrics) to the forwarder. In turn the forwarder will send any waiting new transaction to the backend before stopping (transactions in the retry queue are not retried).

### CODE REVIEW

**Review by ogaca-dd — DISMISSED** [review]
LGTM

File: pkg/forwarder/worker_test.go [hunk 1]
```diff
@@ -102,10 +102,36 @@ func TestWorkerRetryBlockedTransaction(t *testing.T) {
...
+	// to avoid blocking in the Stop method since we don't actually start
+	// the workder
+	w.stopChan = make(chan struct{}, 2)
+	close(w.stopped)
+
+	mockTransaction := newTestTransaction()
+	mockTransaction.On("Process", w.Client).Return(nil).Times(1)
+	mockTransaction.On("GetTarget").Return("").Times(1)
+	highPrio <- mockTransaction
+
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply pkg/forwarder/worker_test.go]: Good idea.

File: releasenotes/notes/flush-data-on-stop-f8a1ab6521076fc9.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+enhancements:
+  - |
+    The agent will now try to flush data to the backend when before exiting
+    (from DogStatsD and checks). This avoid having metrics gap when restarting
+    the agent. This behavior can be disable through configuration, see
+    `aggregator_stop_timeout` and `forwarder_stop_timeou`.
```
Comment (ofek) [inline releasenotes/notes/flush-data-on-stop-f8a1ab6521076fc9.yaml]:
```suggestion
    `aggregator_stop_timeout` and `forwarder_stop_timeout`.
```

**Review by ofek — CHANGES_REQUESTED** [review]

**Review by ofek — DISMISSED** [review]

**Review by truthbk — DISMISSED** [review]
I don't see anything wrong with this, just a few comments. But it LGTM!

**Review by ogaca-dd — DISMISSED** [review]

File: pkg/config/config_template.yaml [hunk 3]
```diff
@@ -131,6 +131,18 @@ api_key:
 #
 # histogram_copy_to_distribution_prefix: "<PREFIX>"
 
+## @param aggregator_stop_timeout - integer - optional - default: 2
+## When stopping the agent, the Aggregator will try to flush out data ready for
+## aggregation (metrics, events, ...). Data are flushed to the Forwarder in order
+## to be sent to DataDog, therefore the Agent might take at most
+## 'aggregator_stop_timeout'+'forwarder_stop_timeout' seconds to exit.
+##
+## You can set the maximum amount of time, in seconds, allocated to the
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply pkg/config/config_template.yaml]: By default systemd has a 90s default timeout for stopping a service. Windows SCM has a default of 30s. So our 4s default if fine. In both case a user increase the agent timeout they can/have to change the default of our service through systemd and Windows.

But I don't think we will encounter the issue.

File: pkg/aggregator/aggregator.go [hunk 4]
```diff
@@ -536,22 +568,39 @@ func (agg *BufferedAggregator) flushEvents(start time.Time) {
...
+func (agg *BufferedAggregator) flush(start time.Time, waitForSerializer bool) {
+	agg.flushSeriesAndSketches(start, waitForSerializer)
+	agg.flushServiceChecks(start, waitForSerializer)
+	agg.flushEvents(start, waitForSerializer)
+}
+
+// Stop stops the aggregator. Based on 'flushData' waiting metrics (from checks
+// or closed dogstatsd buckets) will be sent to the serializer before stopping.
+func (agg *BufferedAggregator) Stop() {
+	agg.stopChan <- struct{}{}
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply pkg/aggregator/aggregator.go]: Yes, the goal is to purge data that are ready.

File: pkg/aggregator/aggregator.go [hunk 5]
```diff
@@ -448,24 +462,27 @@ func (agg *BufferedAggregator) sendSeries(series metrics.Series, start time.Time
...
-func (agg *BufferedAggregator) flushSeriesAndSketches(start time.Time) {
+func (agg *BufferedAggregator) flushSeriesAndSketches(start time.Time, waitForSerializer bool) {
 	series, sketches := agg.GetSeriesAndSketches()
 
-	agg.sendSketches(sketches, start)
-	agg.sendSeries(series, start)
+	// Serialize and forward sketches in a separate goroutine
+	addFlushCount("Sketches", int64(len(sketches)))
+	if len(sketches) != 0 {
+		if waitForSerializer {
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply pkg/aggregator/aggregator.go]: Yes :+1:

**Pull request merged**